provider/plugin/httpapi.go

```diff
@@ -0,0 +1,166 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package plugin
+
+import (
+	"context"
+	"encoding/json"
+	"net"
+	"net/http"
+	"time"
+
+	"sigs.k8s.io/external-dns/endpoint"
+	"sigs.k8s.io/external-dns/plan"
+	"sigs.k8s.io/external-dns/provider"
+
+	log "github.com/sirupsen/logrus"
+)
+
+type HTTPProvider struct {
+	provider provider.Provider
+}
+
+type PropertyValuesEqualsRequest struct {
+	Name     string `json:"name"`
+	Previous string `json:"previous"`
+	Current  string `json:"current"`
+}
+
+type PropertyValuesEqualsResponse struct {
+	Equals bool `json:"equals"`
+}
+
+func (p *HTTPProvider) recordsHandler(w http.ResponseWriter, req *http.Request) {
+	if req.Method == http.MethodGet { // records
+		records, err := p.provider.Records(context.Background())
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		w.Header().Set("Content-Type", "application/json")
+		w.WriteHeader(http.StatusOK)
+		json.NewEncoder(w).Encode(records)
+		return
+	} else if req.Method == http.MethodPost { // applychanges
+		var changes plan.Changes
+		if err := json.NewDecoder(req.Body).Decode(&changes); err != nil {
+			w.WriteHeader(http.StatusBadRequest)
+			return
+		}
+		err := p.provider.ApplyChanges(context.Background(), &changes)
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		w.Header().Set("Content-Type", "application/json")
+		w.WriteHeader(http.StatusOK)
+		return
+	}
+	log.Errorf("Unsupported method %s", req.Method)
+	w.WriteHeader(http.StatusBadRequest)
+}
+
+func (p *HTTPProvider) propertyValuesEqualHandler(w http.ResponseWriter, req *http.Request) {
+	if req.Method != http.MethodPost {
+		log.Errorf("Unsupported method %s", req.Method)
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	pve := PropertyValuesEqualsRequest{}
+	if err := json.NewDecoder(req.Body).Decode(&pve); err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+	b := p.provider.PropertyValuesEqual(pve.Name, pve.Previous, pve.Current)
+	r := PropertyValuesEqualsResponse{
+		Equals: b,
+	}
+	out, err := json.Marshal(&r)
+	if err != nil {
+		log.Error(err)
+	}
+	w.Header().Set("Content-Type", "application/json")
+	w.Write(out)
+}
+
+func (p *HTTPProvider) adjustEndpointsHandler(w http.ResponseWriter, req *http.Request) {
+	if req.Method != http.MethodPost {
+		log.Errorf("Unsupported method %s", req.Method)
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	pve := []*endpoint.Endpoint{}
+	if err := json.NewDecoder(req.Body).Decode(&pve); err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+	pve = p.provider.AdjustEndpoints(pve)
+	out, err := json.Marshal(&pve)
+	if err != nil {
+		log.Error(err)
+	}
+	w.Header().Set("Content-Type", "application/json")
+	w.Write(out)
+}
+
+func (p *HTTPProvider) negotiate(w http.ResponseWriter, req *http.Request) {
+	w.Header().Set(varyHeader, contentTypeHeader)
```

Comment:
Shouldn't you have to check the Accept Header here for the wanted media format?

Reply:
For "check" you mean "set"? This is only used to negotiate and the two headers that are set are according to the definition.

Reply:
ok, maybe i have a wrong understanding how the negotiating works, but according [this example](http://opensource.zalando.com/restful-api-guidelines/#_example_2) the client tells via accept header, which version he wants , and the server answers if he supports that or not. So the *check* would mean check the accept header...

Reply:
Isn't it supposed to be that after the negotiation all subsequent requests should have :
- an Accept header with the value from the negotiation result , for the case that it receives a response body with content, and
- an Content-Type header with same value, for the case that the request body has content ?

Concrete: 

| Client request | Header to set  |
| -------------- | -------------- |
| `Records` | `Accept` |
| `ApplyChanges` | `Content-Type` |
| `PropertyValuesEqual` | `Content-Type` and `Accept` |
| `AdjustEndpoints` | `Content-Type` and `Accept` |

Reply:
AFAIS, current implementation of [`plugin.go`](https://github.com/kubernetes-sigs/external-dns/blob/raffo/plugin-provider/provider/plugin/plugin.go) is setting the headers like above.
So, is resolved FMPOV